src/front/wgsl/index.rs

```diff
@@ -0,0 +1,143 @@
+use super::{ast, Error};
+use crate::{FastHashMap, Handle, Span};
+
+pub struct Index<'a> {
+    dependency_order: Vec<Handle<ast::GlobalDecl<'a>>>,
+}
+
+impl<'a> Index<'a> {
+    pub fn generate(tu: &ast::TranslationUnit<'a>) -> Result<Self, Error<'a>> {
+        let mut globals = FastHashMap::with_capacity_and_hasher(tu.decls.len(), Default::default());
+
+        // Populate dependencies
+        for (handle, decl) in tu.decls.iter() {
+            let ident = decl_ident(decl);
+            let name = ident.name;
+            if let Some(old) = globals.insert(name, handle) {
+                return Err(Error::Redefinition {
+                    previous: decl_ident(&tu.decls[old]).span,
+                    current: ident.span,
+                });
+            }
+        }
+
+        let len = tu.decls.len();
+        let solver = DependencySolver {
+            globals: &globals,
+            module: tu,
+            visited: vec![false; len],
+            temp_visited: vec![false; len],
+            path: Vec::new(),
+            out: Vec::with_capacity(len),
+        };
+        let dependency_order = solver.solve()?;
+
+        Ok(Self { dependency_order })
+    }
+
+    pub fn visit_ordered(&self) -> impl Iterator<Item = Handle<ast::GlobalDecl<'a>>> + '_ {
+        self.dependency_order.iter().copied()
+    }
+}
+
+pub struct ResolvedDependency<'a> {
+    pub decl: Handle<ast::GlobalDecl<'a>>,
+    pub usage: Span,
+}
+
+struct DependencySolver<'source, 'temp> {
+    globals: &'temp FastHashMap<&'source str, Handle<ast::GlobalDecl<'source>>>,
+    module: &'temp ast::TranslationUnit<'source>,
+    visited: Vec<bool>,
+    temp_visited: Vec<bool>,
+    path: Vec<ResolvedDependency<'source>>,
+    out: Vec<Handle<ast::GlobalDecl<'source>>>,
+}
+
+impl<'a> DependencySolver<'a, '_> {
+    fn solve(mut self) -> Result<Vec<Handle<ast::GlobalDecl<'a>>>, Error<'a>> {
+        for id in self.module.decls.iter().map(|x| x.0) {
```

Comment:
The way Naga generally does this is clearer, I think:

```suggestion
        for (id, _decl) in self.module.decls.iter() {
```